curated_transformers/models/attention.py

```diff
@@ -61,12 +70,112 @@ def forward(
         attn_scores = q @ k.transpose(-2, -1)
         attn_scores /= math.sqrt(model_dim)
 
-        # Replace tokens that we don't want to attend to with a large
-        # negative value to zero them out during softmax normalization.
-        batch, seq_len = attn_mask.shape
-        attn_scores += attn_mask.logit_mask.view(batch, 1, 1, seq_len)
+        if attn_mask is not None:
+            # Replace tokens that we don't want to attend to with a large
+            # negative value to zero them out during softmax normalization.
+            batch, seq_len = attn_mask.shape
+            attn_scores += attn_mask.logit_mask.view(batch, 1, 1, seq_len)
 
         attn_weights = attn_scores.softmax(dim=-1)
         attn_values = self.dropout(attn_weights @ v)
 
         return attn_values
+
+
+# https://www.tensorflow.org/text/tutorials/transformer#multi-head_attention
+class SelfAttention(Module):
+    """Transformer self-attention layer (Vaswani et al., 2016)."""
+
+    def __init__(
+        self,
+        *,
+        dropout_prob: float = 0.1,
+        hidden_width: int = 768,
+        num_attention_heads: int = 12,
+    ):
+        """Construct a self-attention layer.
+
+        dropout_prob (float): dropout to apply between the self-attention
+            and output layers (default: 0.1).
+        hidden_width (int): hidden width of the layer (default: 768).
+        num_attention_heads (int): number of attenion heads (default: 12).
+        """
+        super().__init__()
+
+        self.model_dim = hidden_width
+        self.num_heads = num_attention_heads
+        if self.model_dim % self.num_heads != 0:
+            raise ValueError(
+                Errors.E003.format(
+                    hidden_width=self.model_dim, num_heads=self.num_heads
+                )
+            )
+
+        self.dims_per_head = self.model_dim // self.num_heads
+        self.attention = ScaledDotProductAttention(dropout_prob=dropout_prob)
+        self.input = Linear(self.model_dim, self.model_dim * 3)
+        self.output = Linear(self.model_dim, self.model_dim)
+
+    def forward(self, x: Tensor, attn_mask: Optional[AttentionMask]) -> Tensor:
+        """
+        Apply self-attention layer to the input.
+
+        Sequence elements that are marked with `False` in the attention mask
+        are ignored by the attention mechanism (if a mask is provided).
+
+        x (Tensor): input to apply self-attention to.
+        attn_mask (Optional[AttentionMask]): attention mask.
+
+        Shapes:
+            x - (batch, seq_len, width)
+            attn_mask - (batch, seq_len)
+        """
+
+        proj = self.input(x)
+        q, k, v = proj.chunk(3, dim=-1)
```

Comment:
Perhaps a comment here to explain why we're doing the projection just once.

Reply:
Done.